File: code/clustering.jl
using LightGraphs, MetaGraphs, GraphIO
using ParserCombinator #Needed for GML reading

graph_name = "simpledistros.gml"
number_subreddits = 6
infomap_clu = "simpledistros.clu"
write_file = "test.gml"

# Read the graph
g = open(graph_name, "r") do io
    GraphIO.GML.loadgml(io, "graph")
end
g = MetaGraph(g)


# Save GML file
function savegml(io::IO, g)
    println(io,"graph")
    println(io,"[")
    for i = 1:nv(g)
        println(io,"\tnode")
        println(io,"\t[")
        println(io,"\t\tid $i")
        labels = props(g,i)
        for (key, val) in labels
            println(io, "\t\t$key $val")
        end
        println(io,"\t]")
    end
    for e in LightGraphs.edges(g)
        s,t =Tuple(e)
        println(io, "\tedge")
        println(io, "\t[")
        println(io, "\t\tsource $s")
        println(io, "\t\ttarget $t")
        println(io, "\t]")
    end
    println(io, "]")
    return 1
end

function indexesof(x, iter)
    map(y->y[1],filter(y->x==y[0],zip(iter,1:length)))
end

# Louvain algorithm for undirected unweigthed graphs
function louvain_community(g)
    improvements = false
    comunities = 1:nv(g)

    # First phase
    change = true
    while change
        change = false
        m = sum(weights)

        for i =1: nv(g)
            for j in neighbors(g,i)
                comunitynodes = indexesof(comunities[j], comunities)
                sumj = count(i->(i==comunities[j]),comunities)
                
                delta = 0 #TODO 
            end
        end
    end
end


# Read Infomap clustering and tag nodes
println("Infomap clustering...")
for line in eachline(infomap_clu)
    # Ignore comments of clu file
    if startswith(line, "#")
        continue
    end

    x = split(line,r"\s")
    set_prop!(g,parse(Int32,x[1]),:map, x[2])
end

open(write_file, "w") do io
    savegml(io, g)
end

File: code/clustering.py
import networkx as nx
from networkx.algorithms.community import label_propagation_communities, greedy_modularity_communities, asyn_fluidc

graph_name = 'simpledistros.gml'
number_subreddits = 6
infomap_clu = 'simpledistros.clu'

G = nx.read_gml(graph_name)
G = nx.convert_node_labels_to_integers(G,first_label=1)

# Get the label propagation clustering
print('Label propagation clustering...')
lpa_gen = label_propagation_communities(G)
partition = []
for community in lpa_gen:
    part = sorted(community)
    partition.append(part)
partition = tuple(partition)
for v in G:
    for i in range(len(partition)):
        if v in partition[i]:
            G.node[v]['lpa'] = i
            continue

# Get the modularity clustering
print('Modularity clustering...')
comp = greedy_modularity_communities(G)
partition = []
for community in comp:
    part = sorted(community)
    partition.append(part)
for v in G:
    for i in range(len(partition)):
        if v in partition[i]:
            G.node[v]['mod'] = i
            continue

# Get the Infomap clustering
print('Infomap clustering...')
with open(infomap_clu) as f:
    content = f.readlines()
content = content[2:] # remove the first two lines
content = [x.split() for x in content]
content = [(int(x[0]), int(x[1])) for x in content]
for x in content:
    G.node[x[0]]['map'] = x[1]


nx.write_gml(G, graph_name)
